src/Compilers/CSharp/Portable/LanguageVersion.cs

```diff
@@ -121,4 +104,107 @@ internal static ErrorCode GetErrorCode(this LanguageVersion version)
             }
         }
     }
-}
+
+    /// <summary>
+    /// This type is attached to diagnostics for required language version and should only be used
+    /// on such diagnostics, as they are scrapped out by <see cref="CSharpCompilation.GetRequiredLanguageVersion"/>.
+    /// </summary>
+    internal class RequiredLanguageVersion : IMessageSerializable
+    {
+        internal LanguageVersion Version { get; }
+
+        internal RequiredLanguageVersion(LanguageVersion version)
+        {
+            Version = version;
+        }
+
+        public override string ToString()
+        {
+            return Version.ToDisplayString();
+        }
+    }
+
+    public static class LanguageVersionExtensions
+    {
+        /// <summary>
+        /// Displays the version number in the format expected on the command-line (/langver flag).
+        /// For instance, "6", "7", "7.1", "latest".
+        /// </summary>
+        public static string ToDisplayString(this LanguageVersion version)
+        {
+            switch (version)
+            {
+                case LanguageVersion.CSharp1:
+                    return "1";
+                case LanguageVersion.CSharp2:
+                    return "2";
+                case LanguageVersion.CSharp3:
+                    return "3";
+                case LanguageVersion.CSharp4:
+                    return "4";
+                case LanguageVersion.CSharp5:
+                    return "5";
+                case LanguageVersion.CSharp6:
+                    return "6";
+                case LanguageVersion.CSharp7:
+                    return "7";
+                case LanguageVersion.Default:
+                    return "default";
+                case LanguageVersion.Latest:
+                    return "latest";
+                default:
+                    throw ExceptionUtilities.UnexpectedValue(version);
+            }
+        }
+
+        public static bool TryParseDisplayString(this LanguageVersion _, string newVersion, out LanguageVersion result)
```

Comment:
I probably prefer to leave the parsing method as-is (not throwing exception) and just handle bad input like other parsing errors.

---
In reply to: [103990545](https://github.com/dotnet/roslyn/pull/17435#discussion_r103990545) [](ancestors = 103990545)